### PR TITLE
Drag and Drop on Undocked Image Panel - Fix issue #782

### PR DESCRIPTION
**Pull request by AlexanderGetka-cbica**
Fixes issue #782
Drag-n-drop events on the undocked image panel were not being handled.
Now the tool tab dock is a custom QDockWidget derivative class, DragAndDropSharingDockWidget, that emits the drag-and-drop events as signals. fMainWindow connects these to its own drag-and-drop handlers.

Also fixes a minor error in fMainWindow where the tool tabs were not children of the main window, causing the UI style to propagate very strangely.

## Proposed Changes
  - 
  -
  -

### CODE REVIEW

File: src/view/gui/fMainWindow.cpp [hunk 1]
```diff
@@ -252,9 +254,13 @@ fMainWindow::fMainWindow()
   m_tabWidget->setMinimumHeight(minheight);
   m_tabWidget->setMaximumHeight(m_tabWidget->minimumHeight());
 
-  m_toolTabdock = new QDockWidget();
+  m_toolTabdock = new DragAndDropSharingDockWidget(); // custom class to propagate drag-and-drop events to the main window
```
Comment (ashishsingh18) [inline src/view/gui/fMainWindow.cpp]: Please make sure there is no memory leak.

File: src/view/gui/CMakeLists.txt [hunk 2]
```diff
@@ -52,6 +52,7 @@ SET( GUI
   fSBRTAnalysisDialog
   fTexturePipelineDialog
   AppearancePage
+  DragAndDropSharingDockWidget
```
Comment (ashishsingh18) [inline src/view/gui/CMakeLists.txt]: Another name would be better such as "QCustomDockWidget" or "CaptkDockwidget". As of now we are just handling drag and drop, but if in the future we extend it further, the current name will be a misnomer.

Comment (AlexanderGetka-cbica) [reply src/view/gui/CMakeLists.txt]: I just pushed a commit that changes this to CaptkDockWidget.

File: src/view/gui/fMainWindow.cpp (hunk 1 above)
Comment (AlexanderGetka-cbica) [reply src/view/gui/fMainWindow.cpp]: I just added a check and delete for this in fMainWindow's destructor.

Comment (dboun) [reply src/view/gui/fMainWindow.cpp]: @AlexanderGetka-cbica  In Qt, QObjects and derivatives get destroyed when their parent does. In this particular case, it looks like the correct parent is fMainWindow itself. Do

```cpp
m_toolTabdock = new QDockWidget(this);
```

and then you don't have to worry about lifetime management.

(QMainWindow::addDockWidget() might set parent regardless, but whatever)

Comment (AlexanderGetka-cbica) [reply src/view/gui/fMainWindow.cpp]: @dboun I was aware of this and tried to set its parent directly to be fMainWindow. For some reason we couldn't figure out  yesterday, this causes the entire docking tool tab to be hidden on the gui, even if we explicitly call show() on it.

Comment (AlexanderGetka-cbica) [reply src/view/gui/fMainWindow.cpp]: I just discovered that setting the window flag for the dock widget to Qt::Window causes it to be hidden if it's a child of the main window. I set it to Qt::Subwindow and this works fine on my machine running Ubuntu, but the window flags have a platform-dependent effect so it's possible this ends up looking strange on mac/windows.

File: src/view/gui/AppearancePage.cpp [hunk 3]
```diff
@@ -30,20 +30,23 @@ AppearancePage::AppearancePage(QWidget *parent) :
...
+    if (!QVariant(ApplicationPreferences::GetInstance()->GetFileAvailability()).toBool())
 	{
 		ui->themeComboBox->setCurrentIndex(ThemeType::Dark);
         ApplicationPreferences::GetInstance()->SetFont(this->m_SelectedFont.toString());
         ApplicationPreferences::GetInstance()->SetTheme(QVariant::fromValue(this->m_SelectedTheme).toString());
         ApplicationPreferences::GetInstance()->DisplayPreferences();
-	}
+    }
+    //! Apply
+    OnChangeTheme(m_SelectedTheme);
```
Comment (ashishsingh18) [inline src/view/gui/AppearancePage.cpp]: This call may not be needed here.

File: src/view/gui/AppearancePage.cpp [hunk 4]
```diff
@@ -54,6 +57,9 @@ AppearancePage::~AppearancePage()
 
 void AppearancePage::OnChangeTheme(int theme)
 {
+    // OnChangeTheme updates the instance variables to contain the selected theme and style sheet but does not apply it.
+    m_PreviousTheme = m_SelectedTheme;
```
Comment (ashishsingh18) [inline src/view/gui/AppearancePage.cpp]: These 2 lines may not be needed here. We update the previous with selected only when we hit okay.

File: src/view/gui/AppearancePage.cpp [hunk 5]
```diff
@@ -62,7 +68,7 @@ void AppearancePage::OnChangeTheme(int theme)
 #ifdef QT_DEBUG
 		qssPath = getCaPTkDataDir() + /*"../etc/"*/ + "/captk.qss";
 #else
-		qssPath = getCaPTkDataDir() + "../etc/" + CAPTK_STYLESHEET_FILE;
+        qssPath = getCaPTkDataDir() + /*"../etc/"*/ + CAPTK_STYLESHEET_FILE;
```
Comment (ashishsingh18) [inline src/view/gui/AppearancePage.cpp]: Are you sure this works in release?

File: src/view/gui/AppearancePage.cpp [hunk 6]
```diff
@@ -107,54 +113,29 @@ void AppearancePage::OnSelectFontButtonClicked()
 void AppearancePage::OnOkay()
 {
     //! apply selected font and stylesheet for preview
-	qApp->setFont(this->m_SelectedFont);
-	qApp->setStyleSheet(this->m_SelectedStyleSheet);
+    qApp->setFont(m_SelectedFont);
+    OnChangeTheme(m_SelectedTheme);
```
Comment (ashishsingh18) [inline src/view/gui/AppearancePage.cpp]: This may again not be needed here.

File: src/view/gui/AppearancePage.cpp [hunk 7]
```diff
@@ -170,3 +151,55 @@ void AppearancePage::Restore()
     ApplicationPreferences::GetInstance()->DisplayPreferences();
```
Comment (ashishsingh18) [inline src/view/gui/AppearancePage.cpp]: Have you tested the functionality to restore ? I guess this change would have broken it.

File: src/view/gui/AppearancePage.h [hunk 8]
```diff
@@ -46,6 +46,23 @@ public slots:
...
     Ui::AppearancePage *ui;
 
+    /**
+    \brief UndoSelectedChanges
+    */
+    void UndoSelectedChanges();
+    /**
+    \brief RefreshAppToSelectedAppearance
+    */
+    void RefreshAppToSelectedAppearance();
```
Comment (ashishsingh18) [inline src/view/gui/AppearancePage.h]: rename to 'ApplyAppearance'.

File: src/view/gui/AppearancePage.cpp [hunk 9]
```diff
@@ -170,3 +151,55 @@ void AppearancePage::Restore()
...
+                             " font : " + systemDefault + " }";
+    confirmationBox->setText("You are seeing a preview of your selected changes. Do you wish to keep them?");
+    confirmationBox->setInformativeText("If you do not confirm within 10 seconds, these changes will automatically revert.");
+    confirmationBox->setStyleSheet(tempStyleSheet);
+    confirmationBox->setStandardButtons(QMessageBox::Apply | QMessageBox::Cancel);
+    confirmationBox->setIcon(QMessageBox::NoIcon);
+
+    //! wait some time and then return the response value (integer)
+    int timeout_in_milliseconds = 10000; // set a reasonable timeout before we auto-cancel
+    QTimer* confirmationTimer = new QTimer(this);
```
Comment (ashishsingh18) [inline src/view/gui/AppearancePage.cpp]: need to clear allocated memory

**Comment by ashishsingh18** [comment]
@AlexanderGetka-cbica Let's discuss these changes before you actually start on them.

**Pull request closed without merging**